src/pip/_internal/models/format_control.py

```diff
@@ -23,7 +25,16 @@ def __init__(self, no_binary=None, only_binary=None):
 
     def __eq__(self, other):
         # type: (object) -> bool
-        return self.__dict__ == other.__dict__
+        if not isinstance(other, self.__class__):
+            return False
+
+        if self.__slots__ != other.__slots__:
+            return False
```

Comment:
Again, this is suspect to false positives when `__slots__` is redefined.

Example:

```python
class Foo:
    __slots__ = ("a", "b")

    def __init__(self, a, b):
        self.a = a
        self.b = b

    def __eq__(self, other):
        if self.__slots__ != other.__slots__:
            return False
        return all(
            getattr(self, k) == getattr(other, k)
            for k in self.__slots__
        )

class Bar(Foo):
    __slots__ = ("b", "a")

# False!
print(Foo(1, 2) == Bar(1, 2))
```

Reply:
I wonder if we want to be a little more explicit rather than pulling out a bunch of hacks here :smile:  I mean there are only 2 attributes, which can be compared manually.  Subclasses extending this may define their own comparison methods.

Reply:
I'm inclined to leave this as-is for now.

Reply:
We can also do

```python
def __eq__(self, other):
    return sorted(self.__slots__) == sorted(other.__slots__)
```